### PR TITLE
attacher: try different kernel sources if there is none in default locations

### PR DESCRIPTION
**Pull request by rootfs**
- [x] kepler can try different kernel source locations
- [x] add kernel sources to kepler container
- [x] test a standalone kepler with the new option and different kernel source dir
- [x] test kepler image on a k8s env that has no kernel devel
- [x] add doc how to add kernel sources and support custom environment (assigned to @SamYuan1990)

#716

### CODE REVIEW

**Comment by marceloamaral** [comment]
Is this related to the PR #728?

**Comment by rootfs** [comment]
no, this is just to allow bcc to compile against pre-installed kernel source, while #728 is to install pre-compiled ebpf module

File: pkg/bpfassets/attacher/bcc_attacher.go [hunk 1]
```diff
@@ -138,9 +139,24 @@ func AttachBPFAssets() (*BpfModuleTables, error) {
 	}
 	// TODO: verify if ebpf can run in the VM without hardware counter support, if not, we can disable the HC part and only collect the cpu time
 	m, err := loadModule(objProg, options)
+	if err != nil {
+		klog.Infof("failed to attach perf module with options %v: %v, from default kernel source.\n", options, err)
+		dirs := config.GetKernelSourceDir()
```
Comment (jichenjc) [inline pkg/bpfassets/attacher/bcc_attacher.go:144]: nit: better rename it to SourceDirs if we allow multiple dirs ..

File: pkg/config/config.go [hunk 2]
```diff
@@ -87,6 +87,9 @@ var (
 
 	configPath = "/etc/kepler/kepler.config"
 
+	// dir of kernel sources for bcc
+	kernelSourceDir = []string{}
```
Comment (jichenjc) [inline pkg/config/config.go:91]: seems it's a plural ?

File: build/kernel-source-images/Dockerfile.ubi [hunk 3]
```diff
@@ -0,0 +1,4 @@
+FROM ImageName
+
+ARG ARCH=amd64
```
Comment (jichenjc) [inline build/kernel-source-images/Dockerfile.ubi:3]: @SamYuan1990 seems we need add s390x `me too` as well?

File: build/Dockerfile [hunk 4]
```diff
@@ -32,4 +32,10 @@ ADD https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-serv
 
 ADD https://raw.githubusercontent.com/sustainable-computing-io/kepler-model-server/main/tests/test_models/AbsComponentModelWeight/Full/KerasCompWeightFullPipeline/KerasCompWeightFullPipeline.json /var/lib/kepler/data/KerasCompWeightFullPipeline.json
 
+# pre install kernel sources
+RUN mkdir -p /usr/share/kepler/kernel_sources
+
+COPY --from=quay.io/sustainable_computing_io/kepler_kernel_source_images:ubi8 /usr/src/kernels /usr/share/kepler/kernel_sources
```
Comment (jichenjc) [inline build/Dockerfile:38]: will this be kept in the released image? how big it is for the copied files?

**Comment by marceloamaral** [comment]
But If #728 cannot be done, do we still need this PR?

File: pkg/bpfassets/attacher/bcc_attacher.go [hunk 5]
```diff
@@ -138,9 +139,24 @@ func AttachBPFAssets() (*BpfModuleTables, error) {
...
+		for _, dir := range dirs {
+			klog.Infof("try to load eBPF module with kernel source dir %s\n", dir)
+			os.Setenv("BCC_KERNEL_SOURCE", dir)
+			m, err = loadModule(objProg, options)
+			if err != nil {
+				klog.Infof("failed to attach perf module with options %v: %v, from kernel source %q\n", options, err, dir)
+			} else {
+				klog.Infof("Successfully load eBPF module with option: %s from kernel source %q", options, dir)
```
Comment (sthaha) [inline pkg/bpfassets/attacher/bcc_attacher.go:152]: nit: 
```suggestion
				klog.Infof("Successfully loaded eBPF module with option: %s from kernel source %q", options, dir)
```

File: pkg/bpfassets/attacher/bcc_attacher.go [hunk 6]
```diff
@@ -138,9 +139,24 @@ func AttachBPFAssets() (*BpfModuleTables, error) {
 	}
 	// TODO: verify if ebpf can run in the VM without hardware counter support, if not, we can disable the HC part and only collect the cpu time
 	m, err := loadModule(objProg, options)
+	if err != nil {
+		klog.Infof("failed to attach perf module with options %v: %v, from default kernel source.\n", options, err)
+		dirs := config.GetKernelSourceDir()
+		for _, dir := range dirs {
+			klog.Infof("try to load eBPF module with kernel source dir %s\n", dir)
```
Comment (sthaha) [inline pkg/bpfassets/attacher/bcc_attacher.go:146]:
```suggestion
			klog.Infof("trying to load eBPF module with kernel source dir %s\n", dir)
```

File: pkg/config/config.go [hunk 7]
```diff
@@ -148,6 +151,30 @@ func getConfig(configKey, defaultValue string) (result string) {
 	return
 }
 
+// SetKernelSourceDir sets the directory for all kernel source. This is used for bcc. Only the top level directory is needed.
+func SetKernelSourceDir(dir string) {
```
Comment (sthaha) [inline pkg/config/config.go:155]: should this function return an error if the `root` dir is invalid? 

```suggestion
func SetKernelSourceDir(dir string) error {

fileInfo, err := os.Stat(path)
if err != nil {
  return err
}

if !fileInfo.IsDir() {
   return fmt.Errorf("expected  kernel root path %s to be a directory", dir)
}
```

Also how we rename the function to `func SetKernelRootDir`

File: pkg/bpfassets/attacher/bcc_attacher.go (hunk 1 above)
Comment (sthaha) [reply pkg/bpfassets/attacher/bcc_attacher.go:144]: +1 to the suggestion above and if you follow https://go.dev/doc/effective_go#Getters  idiom, the recommended name would be `config.KernelSourceDirs()` 👼

**Comment by rootfs** [comment]
I ran the test on my macbook, the pre installed kernel option passed the test, i can see kepler output from there.

File: build/Dockerfile (hunk 4 above)
Comment (rootfs) [reply build/Dockerfile:38]: each kernel-devel is about 5M

**Comment by rootfs** [comment]
@SamYuan1990 can you add this to the section in [your task](https://github.com/sustainable-computing-io/kepler-doc/issues/59) to document new OS support?

**Comment by rootfs** [comment]
@husky-parul can you add this option to the operator after merge? thanks

**Comment by rootfs** [comment]
tested and review feedback addressed, PTAL, thanks

**Comment by rootfs** [comment]
cc @yellowhat for helm update

**Pull request merged**